Add support for TLS 1.3 record protection routines

hanno-arm opened the pull request:
__Context:__ TLS 1.3 uses a modified record protection mechanism which allows (a) flexible length padding, and (b) hiding of a record's content type. Specifically, prior to encryption a record's plaintext is wrapped into a `TLSInnerPlaintext`:
```
      struct {
          opaque content[TLSPlaintext.length];
          ContentType type;
          uint8 zeros[length_of_padding];
      } TLSInnerPlaintext;
```
which appends the record's real content type and an arbitrarily sized 0-padding. The content type on-the-wire is set to `ApplicationData` regardless of the true content type. 

__Summary:__ This PR enhances Mbed TLS' record protection routines with support for TLS 1.3 record protection and adds unit tests to the SSL test suite to exercise them. Since adding support for TLS 1.3 is ongoing but unfinished, the changes happen under a new configuration flag `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` which shall subsequently be used to guard all TLS 1.3 code until an MVP support for TLS 1.3 is available, and at that point it shall be renamed to `MBEDTLS_SSL_PROTO_TLS1_3` in line with previous protocol version identifiers.

__Detail:__  Mbed TLS already supports the Connection ID extension to DTLS 1.2, which uses the same record encryption mechanism as TLS 1.3, with the only distinction that the outer record content type is hardcoded to be a CID specific type instead of ApplicationData. The PR widens the scope of the corresponding record protection routines accordingly to include TLS 1.3. This way, support for TLS 1.3 record protection can be added with very few actual code changes.

sbutcher-arm commented on library/ssl_msg.c:
```diff
@@ -574,6 +578,28 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
+         * Note that this changes `rec->data_len`, and hence
+         * `post_avail` needs to be recalculated afterwards.
+         */
+        if( ssl_build_inner_plaintext( data,
```
Suggestion: We're calling `ssl_build_inner_plaintext()` here and below. Is it possible for `transform->minor_ver == MBEDTLS_SSL_MINOR_VERSION_4` and `transform->out_cid_len != 0`, in which case would it be better and simpler to combine the two potential calls to `ssl_build_inner_plaintext()`?

hanno-arm replied:
Good point, I wondered about that, too. For now I have deliberately kept the code paths separate because there's no overlap of the TLS 1.3 code and CID code. 

That said, at least a comment saying as much might be helpful.

hanno-arm replied:
Done.

sbutcher-arm commented on library/ssl_msg.c:
```diff
@@ -1549,11 +1575,23 @@ int mbedtls_ssl_decrypt_buf( mbedtls_ssl_context const *ssl,
+    if( transform->minor_ver == MBEDTLS_SSL_MINOR_VERSION_4 )
+    {
+        /* Remove inner padding and infer true content type. */
+        ret = ssl_parse_inner_plaintext( data, &rec->data_len,
```
Suggestion: And again, it it simpler to have one call to `ssl_parse_inner_plaintext()` with a more complex condition?

hanno-arm replied:
I tried and found it a bit difficult to read, so preferred the present solution.

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -1706,6 +1706,22 @@
+ * Uncomment this macro to enable experimental and partial
+ * functionality specific to TLS 1.3.
+ */
+//#define MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
```
Observation: I like `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` having the word `EXPERIMENTAL` in it, as it will deter people from trying it out too soon. However, when we rename it to `MBEDTLS_SSL_PROTO_TLS1_3`, there will be a lot of searching and replacing and fiddling with broken formatting, which will be a really tedious task.

hanno-arm replied:
Hm, I don't follow - it's going to be a straightforward `sed`, or not?

mpg replied:
I agree, replacement is easily done with `sed`. I also agree with Simon that dumb replacement is in general not enough as it can leave the formatting broken (line length, vertical alignment). However in this specific case I think this unlikely to be an issue as most uses are going to be
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
```
on a line on its own. The few instances where it's not the case can by fixed manually if necessary (which is made less likely by the fact that we'd be changing to a shorter name).

So overall, I think the benefits of having `_EXPERIMENTAL` in the name outweigh the potential costs.

hanno-arm replied:
No action here, then.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1706,6 +1706,22 @@
+/**
+ * \def MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+ *
+ * This is an experimental macro used to selectively enable parts
```
Sounds good, I'll change accordingly :+1:

hanno-arm replied:
Changed.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1706,6 +1706,22 @@
+ * a prototype implementation for TLS 1.3 and DTLS 1.3 but aren't
+ * used otherwise.
+ *
+ * \warning Features under the control of this macro are experimental
```
Ok, I'll strengthen the wording.

hanno-arm replied:
Done.

hanno-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3178,6 +3178,26 @@ void ssl_crypt_record( int cipher_type, int hash_id,
             continue;
         }
 
+#if defined(MBEDTLS_SSL_DTLS_CONNECTION_ID)
```
Agreed, though I'd prefer to keep it here.

hanno-arm commented:
@sbutcher-arm @mpg Thank you for your comments, which I believe are addressed now. Could you please re-review?

yanesca commented:
Ah, I just noticed that the CI is failing. Could you please have a look?

And one more thing, could you please use `commit -s` to make the DCO requirement happy?

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1706,6 +1706,25 @@
 #define MBEDTLS_SSL_PROTO_TLS1_2
 
+/**
+ * \def MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
```
@yanesca I'd rather not try to track in detail what has been implemented and what not within the source files, since it complicates matters and is of no benefit to the user IMO insofar as it is about features that aren't yet accessible from the public API.

yanesca replied:
I didn't mean it for the sake of the user, I would like to have that documentation for the maintainer team's sake. The goal of this documentation would be to make it easier to assess where we are, get up to speed if someone new starts reviewing the next bit, or just if there is a long time between reviews.

Again, I wouldn't want to ask for full and detailed architecture description and I don't think that is worth going into excessive details or giving instructions and documentation for hypothetical users.

hanno-arm replied:
@yanesca I'll add a short document where we can keep track of what's already implemented under the umbrella of `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL`.

hanno-arm commented on library/ssl_msg.c:
```diff
@@ -574,6 +578,33 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
+            return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+        }
+
+        rec->type = MBEDTLS_SSL_MSG_APPLICATION_DATA;
```
@yanesca Yes, with TLS 1.3, we have to separate wire-version from logical version since TLS 1.3 records use the TLS 1.2 version field on the wire. In fact, they're indistinguishable from TLS 1.2 records.

There's no single solution in how this should be reflected in the values of `rec->version`. There are at least the following options:
* Always use the on-the-wire version byte for `rec->version`, and don't modify it within the record protection routines. This will require some logical-to-wire version conversion setting up an outgoing record prior to encryption. For incoming records, one could in theory correct the version byte `rec->version` after record decryption, but it's of no use since the real protocol version is always taken from `transform->minor_ver`. 
* Switch to/from wire/logical version for `rec->version` in the record protection routines. This would strengthen the encapsulation slightly, but on the other hand, it introduces code elsewhere since in contrast to the content type, there's no hiding of the real version that's specific to encrypted records: Even the initial ClientHello / ServerHello wouldn't use a TLS 1.3 version identifier in the record header. So, if we'd set `rec->version` to be the logical version prior to calling `ssl_encrypt_buf()` - which so far only happens for non-NULL transforms - then we'd need to change something to make sure we don't write the wrong version to the wire.

My inclination is to go with the first option, and to use `rec->version` as a functionally irrelevant copy of the wire-version. In this case, there's nothing to necessarily to change in this PR, because `rec->version` can be anything - it has no impact on record protection. Only when we start using the TLS 1.3 protection from the SSL layer, we have to set it correctly to get the right wire-format.

yanesca replied:
I agree that conceptually it is best to cleanly separate the logical and the wire version. I wasn't looking for a final solution here, for something correct, that we can refactor later if needed. 

If we are willing to put more thought into it, then I think I would prefer to do it in the conversion functions. Converting between them is straightforward and is done by `mbedtls_ssl_read/write_version()`. The problem is that in TLS 1.3, we have two kinds of wire versions: `version` and `legacy_version`. To resolve this we could have separate functions for each. Since `version` is used only in the supported versions extension, it could be safe to change `mbedtls_ssl_write_version()` to write 0x0303 for TLS 1.3 too. (`mbedtls_ssl_read_version()` does not need to change, because in TLS 1.3 the legacy version must be ignored.)

Of course we will need to add a second pair of functions for this conversion when we are writing the supported versions extension. We should document the functions to make this job for our future selves easier.

What do you think?

hanno-arm replied:
Yes, I agree, and I've got something along those lines in preparation, perhaps it could be one of the next chunks to upstream. However, I don't think it belongs to this PR, because it's about writing/reading to/from the wire, while the PR is about the record conversion to/from unprotected/protected.

yanesca replied:
Ok, let's do it in one of the next chunks then.

hanno-arm commented:
@yanesca 

> Also, the additional data and the IV calculation is different in TLS 1.3. Am I assuming it correctly that you want to make those changes in a follow-up PR?

No, that would belong to the scope of this PR and should be fixed - I'll check. Thanks for bringing it up :+1:

hanno-arm commented:
@yanesca @mpg 

As @yanesca mentioned, the bulk of the work was actually missing here, namely adapting AAD and nonce computations. This should now be addressed - could you please take a look at the approach and let me know what you think?

There is an informational RFC with TLS 1.3 test vectors which I'd like to integrate into the SSL test suite to make sure things are correct now, which is still missing. I'll also create a short document stating what TLS 1.3 specific changes the new experimental macro already entails, as suggested by @yanesca.

hanno-arm commented:
Rebased on top of `development` after seeing CI failures which seemed unrelated to the PR.

hanno-arm commented:
@yanesca @mpg c9c7686 adds `docs/architecture/tls13-experimental.md` which briefly explains the purpose of `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` again and can be used to keep track of which functionality is already introduced through the macro.

yanesca commented:
@sbutcher-arm do you think you will have the time to re-review this PR?

sbutcher-arm commented on library/ssl_msg.c:
```diff
@@ -4866,6 +5001,15 @@ int mbedtls_ssl_parse_change_cipher_spec( mbedtls_ssl_context *ssl )
  */
 
+static size_t ssl_transform_get_explicit_iv_len(
+                        mbedtls_ssl_transform const *transform )
```
Style - Isn't the alignment out here?

hanno-arm replied:
This, too, is because the line would be too long otherwise.

sbutcher-arm commented on tests/suites/test_suite_ssl.data:
```diff
@@ -4154,6 +4154,10 @@ Record crypt, AES-128-GCM, 1.2
 
+Record crypt, AES-128-GCM, 1.3
+depends_on:MBEDTLS_AES_C:MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL:MBEDTLS_GCM_C
+ssl_crypt_record:MBEDTLS_CIPHER_AES_128_GCM:MBEDTLS_MD_MD5:0:0:MBEDTLS_SSL_MINOR_VERSION_4:0:0
```
Suggestion - `MBEDTLS_MD_MD5`? GSM is AEAD and checking the source I don't think is used, so if we must pass a nonsense parameter, wouldn't `MBEDTLS_MD_NONE` be better? Using MD5 looks wrong, and made me check the source afterall.

hanno-arm replied:
You're right. On the other hand, this is a fault in the entire SSL record checking test suite, and I'd prefer to fix it in a separate PR than to blow up this one (at least judging from cheer line count). OK?

sbutcher-arm replied:
Agreed. PR's that are too big are so hard to get merged.

hanno-arm commented on docs/architecture/tls13-experimental.md:
```diff
@@ -0,0 +1,36 @@
+Experimental TLS 1.3 develpoments
```
Thank you, will change :+1:

hanno-arm commented on docs/architecture/tls13-experimental.md:
```diff
@@ -0,0 +1,36 @@
+Experimental TLS 1.3 develpoments
```
Will fix

hanno-arm commented on docs/architecture/tls13-experimental.md:
```diff
@@ -0,0 +1,36 @@
+Status
+------
+
+The following lists which parts of the TLS 1.3 prototype have already been upstreamed
```
Hm, good question. It's a bit different in that it's not guarded by `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` because it makes sense to be used on its own. Still it should be mentioned. I will add something.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -3535,6 +3554,22 @@
  */
 //#define MBEDTLS_SSL_CID_PADDING_GRANULARITY 16
 
+/** \def MBEDTLS_SSL_TLS13_PADDING_GRANULARITY
```
The precedent is the class of macros `MBEDTLS_SSL_PROTO_TLS1_X`. I found the current form easier to read, but since @gbryant-arm independently raised the issue and found it confusing, I'll change it.

hanno-arm replied:
Done.

hanno-arm commented on include/mbedtls/config.h:
```diff
@@ -1706,6 +1706,25 @@
+ * Uncomment this macro to enable experimental and partial
+ * functionality specific to TLS 1.3.
+ */
+//#define MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
```
Ok. Right now, there's no functional dependency on HKDF, but eventually there will be, and I don't see harm in listing that already now.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_msg.c:
```diff
@@ -704,52 +810,51 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
                                   iv, transform->ivlen );
         MBEDTLS_SSL_DEBUG_BUF( 4, "IV used (transmitted)",
-                                  data - explicit_iv_len, explicit_iv_len );
+                               data - dynamic_iv_len * dynamic_iv_is_explicit,
```
This is because the line would be too long otherwise.

hanno-arm commented:
@sbutcher-arm Many thanks for your review and comments! I believe to have addressed them. Could you please take another look?

yanesca commented on library/ssl_msg.c:
```diff
@@ -536,76 +536,39 @@ static void ssl_mac( mbedtls_md_context_t *md_ctx,
- *     dst_nonce_len == fixed_iv_len &&
- *     dynamic_iv_len < dst_nonce
+
+/* Compute IV := ( fixed_iv || 0 ) XOR ( 0 || dynamic_IV )
```
This implies the constraints on the lengths of the parameters, but it still might worth noting that the function assumes that `dst_iv_len >= fixed_iv_len` and `dst_iv_len >= dynamic_iv_len`. These are needed for safe operation.

hanno-arm replied:
I've documented the precondition.

yanesca commented on library/ssl_msg.c:
```diff
@@ -536,76 +536,39 @@ static void ssl_mac( mbedtls_md_context_t *md_ctx,
+
+/* Compute IV := ( fixed_iv || 0 ) XOR ( 0 || dynamic_IV )
+ *
+ * Concretely, this occurs in two variants:
```
I think this is a very clever optimisation and I like it very much. But I think the more clever a solution is, the better we need to document it for the sake of maintainability. It might seem overkill, but could we perhaps elaborate on these cases (which versions and ciphers use it, reference to sections in the corresponding RFCs).

hanno-arm replied:
I've added a comment - however, I prefer not to be too verbose here since there are already similar comments when the function is called, as well as in the documentation of `mbedtls_ssl_transform`.

yanesca commented on library/ssl_msg.c:
```diff
@@ -911,7 +868,7 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
          * Prefix record content with dynamic IV in case it is explicit.
          */
-        if( dynamic_iv_is_explicit == 1 )
+        if( dynamic_iv_is_explicit )
```
Minor, style:
```suggestion
        if( dynamic_iv_is_explicit == 1 )
```

hanno-arm replied:
Fixed.

hanno-arm commented on library/ssl_msg.c:
```diff
@@ -805,7 +880,8 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
         MBEDTLS_SSL_DEBUG_BUF( 4, "IV used (transmitted)",
-                                  data - explicit_iv_len, explicit_iv_len );
+                               data - dynamic_iv_len * dynamic_iv_is_explicit,
+                               dynamic_iv_len * dynamic_iv_is_explicit );
```
Well spotted! This certainly needs fixing.

hanno-arm replied:
Fixed.

yanesca replied:
I am sorry I can't see this being fixed in the new commits. Could you please double check that you have pushed the fix for this?

hanno-arm replied:
The commit is https://github.com/ARMmbed/mbedtls/pull/3315/commits/16bf0e234663936ff533292ef64fba7001f694dc Note that the source buffer changed.

hanno-arm commented on library/ssl_msg.c:
```diff
@@ -805,7 +880,8 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
                                   iv, transform->ivlen );
         MBEDTLS_SSL_DEBUG_BUF( 4, "IV used (transmitted)",
-                                  data - explicit_iv_len, explicit_iv_len );
+                               data - dynamic_iv_len * dynamic_iv_is_explicit,
```
I went for a middleway, see the recent commits.

hanno-arm commented:
@yanesca Thank you very much for your review. I believe to have addressed your comments. Could you please re-review?

@sbutcher-arm Could you also re-review the latest commits since your last approval, please?
